<a href="https://colab.research.google.com/github/StevenBryceLee/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Copy_of_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [3]:
### YOUR WORK HERE
import pandas as pd
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
columns = ['symboling','norm_loss','make','fuel','aspiration','doors',
           'bod_style','drv_wheels','eng_loc','wheel_base','length','width',
           'height','curb_weight','engine','cylinders','engine_size',
           'fuel_system','bore','stroke','compression','hp','peak_rpm',
           'city_mpg','hgwy_mpg','price']
names = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.names'
auto = pd.read_csv(url, names=columns, na_values = '?')
auto.head()

symboling  norm_loss         make  ... city_mpg hgwy_mpg    price
0          3        NaN  alfa-romero  ...       21       27  13495.0
1          3        NaN  alfa-romero  ...       21       27  16500.0
2          1        NaN  alfa-romero  ...       19       26  16500.0
3          2      164.0         audi  ...       24       30  13950.0
4          2      164.0         audi  ...       18       22  17450.0

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [6]:
### YOUR WORK HERE
auto.sample(10,random_state=30)

symboling  norm_loss           make  ... city_mpg hgwy_mpg    price
121          1      154.0       plymouth  ...       31       38   6692.0
176         -1       65.0         toyota  ...       27       32  10898.0
24           1      148.0          dodge  ...       31       38   6229.0
88          -1      137.0     mitsubishi  ...       23       30   9279.0
72           3      142.0  mercedes-benz  ...       16       18  35056.0
103          0      108.0         nissan  ...       19       25  13499.0
26           1      148.0          dodge  ...       31       38   7609.0
20           0       81.0      chevrolet  ...       38       43   6575.0
155          0       91.0         toyota  ...       27       32   8778.0
185          2       94.0     volkswagen  ...       27       34   8195.0

[10 rows x 26 columns]

In [7]:
#Is your sample reflective of the population value in regards to highway-mpg? 
import numpy as np

In [8]:
#Find the mean for highway-mpg for the entire dataset
pop_mean = np.mean(auto['hgwy_mpg'])
pop_mean

30.75121951219512

In [9]:
#compare your sample mean estimate to that value. Can you say that they are different?
sample_mean = np.mean(auto.sample(10,random_state=30)['hgwy_mpg'])
sample_mean

32.8

In [ ]:
#I would say this is reflective, with an error

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [10]:
### YOUR WORK HERE
#We know that sample_mean is about 33 mpg on the highway, so this is not true
from scipy import stats

In [13]:
sample = auto.sample(10,random_state=30)

In [14]:
stats.ttest_1samp(sample['hgwy_mpg'], 35, axis=0, nan_policy='propagate')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

In [ ]:
#p value is 0.36. No statistically significant difference seen, therefore we believe the salesman
#We do not reject the null hypothesis

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [16]:
sample2 = auto.sample(100,random_state=30)
stats.ttest_1samp(sample2['hgwy_mpg'], 35, axis=0, nan_policy='propagate')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

In [1]:
#p-value is now very close to 0 ( <0.05 ), therefore we reject the salesman's claim
# Based on the t statistic of -6.52 and the p value of < 0.01, we reject the null hypothesis

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

We have a larger sample, giving us more data about the true population mean. Our smaller sample of 10 does not give us enough information to definitively reject the null hypothesis, or demonstrate a difference between the sample value and the salesman's value. The larger sample of 100 increases our precision of the true mean estimate

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [19]:
### YOUR WORK HERE
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'
cols = [
        'Party',
        'handicapped-infants',
        'water-project-cost-sharing',
        'adoption-of-the-budget-resolution',
        'physician-fee-freeze',
        'el-salvador-aid',
        'religious-groups-in-schools',
        'anti-satellite-test-ban',
        'aid-to-nicaraguan-contras',
        'mx-missile',
        'immigration',
        'synfuels-corporation-cutback',
        'education-spending',
        'superfund-right-to-sue',
        'crime',
        'duty-free-exports',
        'export-administration-act-south-africa',
        ]
votes = pd.read_csv(url,names = cols,na_values='?')
votes.head()

Party  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                    NaN
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [20]:
votes.replace({'n':int(0),'y':int(1)},inplace=True)
votes.head()
#Nans in columns are coercing vals to be floats

Party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [22]:
reps = votes[votes['Party']=='republican']
reps.head()

Party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [23]:
dems = votes[votes['Party']=='democrat']
dems.head()

Party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [33]:
#Find an issue that democrats support more than republicans with p < 0.01 
dem_sat = dems['anti-satellite-test-ban']
rep_sat = reps['anti-satellite-test-ban']
stats.ttest_ind(dem_sat, rep_sat,nan_policy='omit')

Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)

In [29]:
print('dems: ', str(np.mean(dem_sat)))
print('reps: ', str(np.mean(rep_sat)))

dems:  0.7722007722007722
reps:  0.24074074074074073


Conclusion: Based on a t-statistic of 12.52 and a p value of < 0.01, we reject the null hypothesis that there is no difference between democrat and republican voting on the anti-satellite-test-ban issue. Based on the positive t-statistic, as well as examining the means of the two data sets, we conclude that democrats support the anti satellite test ban more than republicans with 99% confidence

In [47]:
#Find an issue that republicans support more than democrats with p < 0.01
dem_edu = dems['education-spending']
rep_edu = reps['education-spending']
stats.ttest_ind(dem_edu, rep_edu,nan_policy='omit')

Ttest_indResult(statistic=-20.500685724563073, pvalue=1.8834203990450192e-64)

In [48]:
print('dems: ', str(np.mean(dem_edu)))
print('reps: ', str(np.mean(rep_edu)))

dems:  0.14457831325301204
reps:  0.8709677419354839


Conclusion: Based on the t statistic of -20.50 and the p-value of < 0.01, we reject the null hypothesis that there is no difference between democrat and republican voting on the education spending issue. Based on the positive t-statistic, as well as examining the means of the two data sets, we conclude that democrats support the anti satellite test ban more than republicans with 99% confidence

In [42]:
dems.columns[1:]

Index(['handicapped-infants', 'water-project-cost-sharing',
       'adoption-of-the-budget-resolution', 'physician-fee-freeze',
       'el-salvador-aid', 'religious-groups-in-schools',
       'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
       'immigration', 'synfuels-corporation-cutback', 'education-spending',
       'superfund-right-to-sue', 'crime', 'duty-free-exports',
       'export-administration-act-south-africa'],
      dtype='object')

In [49]:
#Find an issue where the difference between republicans and democrats has p > 0.1
#Too many issues, and I don't pay enough attention to politics to know offhand
for col in dems.columns[1:]:
  dem_col = dems[col]
  rep_col = reps[col]
  p_val = stats.ttest_ind(dem_col, rep_col,nan_policy='omit')[1]
  if(p_val > 0.1):
    print('Issue: ',col)
    print('P value: ', p_val)
    print('T stat: ', stats.ttest_ind(dem_col, rep_col,nan_policy='omit')[0])
    print('dems: ', str(np.mean(dem_col)))
    print('reps: ', str(np.mean(rep_col)))

Issue:  water-project-cost-sharing
P value:  0.9291556823993485
T stat:  -0.08896538137868286
dems:  0.502092050209205
reps:  0.5067567567567568


In [ ]:
#conclusions

For the issue of water project cost sharing, based on the t statistic of -0.089 and p-value of 0.93 we fail to reject the null hypothesis, and find no significant difference between Democrat and republican voting records.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [52]:
### YOUR WORK HERE
#This was done up above. Now changing for all significant differences for the
#first 5 columns so the output isn't too long
for col in dems.columns[1:6]:
  dem_col = dems[col]
  rep_col = reps[col]
  p_val = stats.ttest_ind(dem_col, rep_col,nan_policy='omit')[1]
  if(p_val < 0.1):
    print('\nIssue: ',col)
    print('P value: ', p_val)
    print('dems: ', str(np.mean(dem_col)))
    print('reps: ', str(np.mean(rep_col)))


Issue:  handicapped-infants
P value:  1.613440327937243e-18
dems:  0.6046511627906976
reps:  0.18787878787878787

Issue:  adoption-of-the-budget-resolution
P value:  2.0703402795404463e-77
dems:  0.8884615384615384
reps:  0.13414634146341464

Issue:  physician-fee-freeze
P value:  1.994262314074344e-177
dems:  0.05405405405405406
reps:  0.9878787878787879

Issue:  el-salvador-aid
P value:  5.600520111729011e-68
dems:  0.21568627450980393
reps:  0.9515151515151515




### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





Is the actual mean of the democratic voting sample different than random behavior on the subject of water project cost sharing?


In [53]:
### YOUR WORK HERE
col = 'water-project-cost-sharing'
sample_mean = np.mean(dems[col])
hypothesized_mean = 0.5
sample_stdev = np.std(dems[col])
sample_size = len(dems[col])

In [54]:
t = (sample_mean - hypothesized_mean) / (sample_stdev / np.sqrt(sample_size))
t

0.0683693626407257

In [58]:
degrees_freedom = len(dems[col]) - 1
degrees_freedom

266

In [55]:
#Check with scipy. How are NaNs handled?
stats.ttest_1samp(dems[col], hypothesized_mean, axis=0, nan_policy='omit')

Ttest_1sampResult(statistic=0.06454972243678961, pvalue=0.9485867005339235)

In [ ]:
#Similar but not the same. I suspect the formulas are a little different under the hood

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [62]:
### YOUR WORK HERE
sample_mean_dems = np.mean(dems[col])
sample_mean_reps = np.mean(reps[col])
st_dev_dems = np.std(dems[col])
st_dev_reps = np.std(reps[col])
sample_size_dems = len(dems[col])
sample_size_reps = len(reps[col])

t = ((sample_mean_dems - sample_mean_reps) / 
     (np.sqrt( (st_dev_dems / sample_size_dems) + 
              (st_dev_reps / sample_size_reps) )))

t

-0.06699121793884157

In [63]:
stats.ttest_ind(dems[col], reps[col],nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

In [64]:
#again similar, but not exactly the same
#ttest_ind just calls ttest_ind_from_stats
#ttest_ind_from_stats calls _ttest_finish
#Which finally returns 
'''prob = distributions.t.sf(np.abs(t), df) * 2  # use np.abs to get upper tail
    if t.ndim == 0:
        t = t[()]

    return t, prob'''
#Apply the t-test to this data (with the assumption that the population variances are equal):

'prob = distributions.t.sf(np.abs(t), df) * 2  # use np.abs to get upper tail\n    if t.ndim == 0:\n        t = t[()]\n\n    return t, prob'

In [66]:
#What happens if the standard deviation changes to equalize population variances?
t = ((sample_mean_dems - sample_mean_reps) / 
     (np.sqrt( (st_dev_reps / sample_size_dems) + 
              (st_dev_reps / sample_size_reps) )))

t

-0.06699228602152178

In [67]:
#Trying to drop NaNs manually. I think np drops NaNs differently
dems_dropped = dems[col].dropna()
reps_dropped = reps[col].dropna()
sample_mean_dems = np.mean(dems_dropped)
sample_mean_reps = np.mean(reps_dropped)
st_dev_dems = np.std(dems_dropped)
st_dev_reps = np.std(reps_dropped)
sample_size_dems = len(dems_dropped)
sample_size_reps = len(reps_dropped)

t = ((sample_mean_dems - sample_mean_reps) / 
     (np.sqrt( (st_dev_dems / sample_size_dems) + 
              (st_dev_reps / sample_size_reps) )))

t

-0.06307055082552462

In [68]:
stats.ttest_ind(dems[col], reps[col],nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

In [69]:
stats.ttest_ind(dems_dropped, reps_dropped,nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823994811)

In [70]:
#Trying to drop NaNs manually. I think np drops NaNs differently
dems_dropped = dems[col].dropna()
reps_dropped = reps[col].dropna()
sample_mean_dems = np.mean(dems_dropped)
sample_mean_reps = np.mean(reps_dropped)
st_dev_dems = np.std(dems_dropped)
st_dev_reps = np.std(reps_dropped)
sample_size_dems = len(dems_dropped)
sample_size_reps = len(reps_dropped)

t = ((sample_mean_dems - sample_mean_reps) / 
     (np.sqrt( ( (st_dev_dems ** 2) / sample_size_dems) + 
              ((st_dev_reps ** 2) / sample_size_reps) )))

t

-0.08919789262856036

In [ ]:
#The problem was that I wasn't squaring the standard deviations. 
#Now I'm getting a fairly close value